server/analyzer/init.go

```diff
@@ -86,6 +91,49 @@ func Init() {
 		analyzer.Rule{Id: ruleId_AddDomainConstraintsToCasts, Apply: AddDomainConstraintsToCasts},
 		analyzer.Rule{Id: ruleId_ReplaceNode, Apply: ReplaceNode},
 		analyzer.Rule{Id: ruleId_InsertContextRootFinalizer, Apply: InsertContextRootFinalizer})
+
+	initEngine()
+}
+
+func initEngine() {
+	plan.ValidateForeignKeyDefinition = validateForeignKeyDefinition
+}
+
+// validateForeignKeyDefinition validates that the given foreign key definition is valid for creation
+func validateForeignKeyDefinition(ctx *sql.Context, fkDef sql.ForeignKeyConstraint, cols map[string]*sql.Column, parentCols map[string]*sql.Column) error {
+	// TODO: this check is too permissive, we should be doing some type checks here
+	for i := range fkDef.Columns {
+		col := cols[strings.ToLower(fkDef.Columns[i])]
+		parentCol := parentCols[strings.ToLower(fkDef.ParentColumns[i])]
+		if !foreignKeyComparableTypes(ctx, col.Type, parentCol.Type) {
+			return sql.ErrForeignKeyColumnTypeMismatch.New(fkDef.Columns[i], fkDef.ParentColumns[i])
+		}
+	}
+	return nil
+}
+
+// foreignKeyComparableTypes returns whether the two given types are able to be used as parent/child columns in a
+// foreign key.
+func foreignKeyComparableTypes(ctx *sql.Context, type1 sql.Type, type2 sql.Type) bool {
+	if !type1.Equals(type2) {
+		// There seems to be a special case where CHAR/VARCHAR/BINARY/VARBINARY can have unequal lengths.
+		// Have not tested every type nor combination, but this seems specific to those 4 types.
+		if type1.Type() == type2.Type() {
+			switch type1.Type() {
```

Comment:
Adding on to my other comment, we aren't too strict on ensuring that these are returned correctly, considering they don't map to all of our types so it's not really possible to do so. Instead, it's better to check for these specific types. You could even use `type1.ID.TypeName()` and switch on the string name. But like `oid` returns `sqltypes.VarChar` cause there's no `oid` type in MySQL so we're just returning _something_.

I think `switch type1.ID` with `case pgtypes.Varchar.ID` should work, but if not you can grab the case values using `Id.CaseString()` and match on the string directly (so you'd just use `case "CaseString_Output"`).